### ISSUE TITLE
[RFC] Retire JDBC bridge

### ISSUE THREAD

**Issue by zhicwu** [ClickHouse/ClickHouse#45052]
Can we please retire JDBC bridge and archive `clickhouse-jdbc-bridge` repo? IMO, it's not suitable for production usage due to below issues:
* Design
  - too many requests for one single query (requests like ping and type inferring should not be mandatory)
  - not designed to support query pass-thru and multiple data formats
  - not suitable for production usage due to lack of key features like access control
  - in theory it can be replaced by URL table function/engine
* Implementation
  - server issues like #36385 and #33240
  - clickhouse-jdbc-bridge is just a hack of existing protocol to support query pass-thru
* Usability - hard to use as it requires configuration on both server and external process (clickhouse-jdbc-bridge)
* Misc - not under active maintenance

Suggest to consider below items(as replacement/successor of jdbc table function and engine) moving forward:
* ClickHouse
   - either re-implement `jdbc` by extending `url`
   - or drop `jdbc` table function & engine and enhance `url` by passing execution context in http headers(as mentioned in ClickHouse/ClickHouse#33803)
* Language Neutral Bridge Server
  Taking Java implementation as an example, I plan to add a new module `clickhouse-bridge-server` in `clickhouse-jdbc` repo as replacement of `clickhouse-jdbc-bridge`(or a personal repo if there's any concern). It's basically a web server passing encoded queries along with execution context(user, server address & query_id etc.) from ClickHouse to external datasource. I may have to add extended SQL grammar(for passing execution context and avoiding quote escaping) in JDBC driver to make it work. Below is an immature example to demonstrate the idea:
  ```sql
  -- JDBC driver will convert the first query to the second for execution
  select * from @url(<datasource id(s)>,<parameters>) {
    -- hides boring stuff like URL encoding and quote escaping
    select table1.col1 from table1 inner join table2 on table1.col2=table2.col2 where table1.col3='a' order by table1.col1
  }
  
  select * from url('http(s)://<clickhouse-bridge-server>/query/<encoded and shortened string representing the query and execution context>')
  ```

**Comment by filimonov** [comment]
Also if you just reimplement it as HTTP - who will serve the HTTP queries? How should they be passed to jdbc driver?

**Comment by filimonov** [comment]
> Taking Java implementation as an example, I plan to add a new module clickhouse-bridge-server in clickhouse-jdbc repo as replacement of clickhouse-jdbc-bridge(or a personal repo if there's any concern). It's basically a web server passing encoded queries along with execution context(user, server address & query_id etc.) from ClickHouse to external datasource. 

May be you can refactor (or rewrite completely bad parts) of jdbc-rbidge instead? Why introducing more entities?

**Comment by zhicwu** [comment]
Thanks for the feedback @filimonov. Sorry I didn't make it clear in the first place, it's not throwing jdbc bridge away without providing alternatives :) As far as I know, `jdbc` table function & engine(HTTP + RowBinary) is merely a subset of `url`. That's why I suggested to either drop `jdbc` or re-implement it by extending `url`(e.g. as an alias or so).

> I know a lot of JDBC-bridge users. For sure, they will not be happy about the idea of removing it.

I'll try to make `clickhouse-bridge-server` backward compatible in case user still need jdbc bridge.

> `<encoded and shortened string representing the query and execution context>`
> Should user do that encoding & shortening? :)

It depends. If someone can help with the server side change by introducing extended grammar(see above example), it's transparent to user and all ClickHouse clients will be supported. In worst case, without any support from server side, I can hide the details in JDBC driver(via SQL rewrite), but this means user has to deal with encoding etc. manually if they use a different client other than JDBC.

> Also if you just reimplement it as HTTP - who will serve the HTTP queries? How should they be passed to jdbc driver?

`clickhouse-bridge-server` will serve requests from both `jdbc` and `url` table functions. Let me explain how this works:
1. user sends `select * from @url(...) { <native query> }` to jdbc driver
2. jdbc driver rewrites the query to `select * from url(<encoded url with both native query and context>)` and then sends to CH
3. CH sends http request accordingly to `clickhouse-bridge-server` for execution
4. `clickhouse-bridge-server` takes care of logging(for auditing purpose), ACL check, user mapping(e.g. CH user to external database user), and streaming between CH and the external datasource

> May be you can refactor (or rewrite completely bad parts) of jdbc-rbidge instead? Why introducing more entities? Why make revolution instead of evolution?

There's no evolution in the past year, so now I believe revolution more ;) Joke aside, as I mentioned in above, `clickhouse-jdbc-bridge` is just a hack of current design with limitations, it's much more easier start over by using generic implementation(with `url` table function) to take it to next level.

Lastly, on a second thought, to avoid confusion at user's end, I'll start with a personal repo to unblock my work and hopefully it can better demonstrate my idea.

**Comment by filimonov** [comment]
> is merely a subset of `url`. 

It's definitely more than just url. Additionally it can start the bridge (if is local), check the connection, encode the query, decide which condition can be pushed down from the external query, etc.

>>   <encoded and shortened string representing the query and execution context>
> > Should user do that encoding & shortening? :)
>
> It depends. If someone can help with the server side change by introducing extended grammar(see above example), i

See https://github.com/ClickHouse/ClickHouse/issues/19320


> user sends select * from @url(...) { <native query> } to jdbc driver

In many cases you don't need query more complex than just SELECT * FROM ... WHERE ts BETWEEN ... 

And it it much simpler to push down those WHERE conditions to inner query if it is a simple table reference (comparing to another query). 

> I'll try to make clickhouse-bridge-server backward compatible in case user still need jdbc bridge.
> clickhouse-bridge-server will serve requests from both

First do it better than clickhouse-jdbc-bridge, after that discussion about removing / obsoleting current bring makes sense :P 

> clickhouse-jdbc-bridge is just a hack of current design with limitations,

The same pattern is used for odbc, for libraries, and potentially for catboost.

**Comment by UnamedRus** [comment]
> If you see design pitfalls / limitations of that - it better to address them for all those brigdes, not trying to improve jdbc only.

It's time to discuss, can we have something better(faster/more features) than plain HTTP?) (ie grpc or something like that)

**Comment by alexey-milovidov** [comment]
I want to include the JDBC bridge in the default ClickHouse server package, similar to the ODBC bridge, to make it more accessible to the users.

Caveat: it should be somehow packaged to a self-contained binary to avoid installing JDK (I hate installing anything if it's not ClickHouse).

**Comment by alexey-milovidov** [comment]
I'm not obsessed with the performance of the JDBC bridge... As long as it can connect to Java databases, it is fine.

**Comment by Slach** [comment]
try to look to graalvm native binaries

**Comment by zhicwu** [comment]
Thanks all for the comments!

> In many cases you don't need query more complex than just SELECT * FROM ... WHERE ts BETWEEN ...
> And it it much simpler to push down those WHERE conditions to inner query if it is a simple table reference (comparing to another query).

Thanks for all the advices @filimonov. Unfortunately, it's not our case. We're facing a large, distributed system evolved for decades. Behind the scene, there are various databases deployed world-wide serving for different purposes, and the number is climbing due to acquisition and more and more micro-services. There's centralized multi-layer DW but it's not agile enough for fast-growing business, not to mention transactional process and headache of latency. As a result, we ended up with aggregation APIs and distributed queries against multiple datasources. We know each of the databases very well and because of the nature of problem complexity, push-down won't help.

> It's time to discuss, can we have something better(faster/more features) than plain HTTP?) (ie grpc or something like that)

@UnamedRus, it does make sense to support multiple protocols. My rough idea is start with `url` function using http/https first, and then later support tcp by leveraging `remote` function.

> I want to include the JDBC bridge in the default ClickHouse server package, similar to the ODBC bridge, to make it more accessible to the users.
> I'm not obsessed with the performance of the JDBC bridge... As long as it can connect to Java databases, it is fine.

@alexey-milovidov, yes, it's clear as we talked about this years ago. I think the misunderstanding of JDBC bridge here is that it does not talk to external database directly, it has to go thru a JDBC driver. This means we have to use GraalVM to convert not only JDBC bridge, but also cherry-picked JDBC drivers, to a large binary. Apart from that, each time you want to add a new JDBC driver or even upgrade driver, you have to re-compile...

Anyway, I'll try to satisfy both sides - a nice-to-have JDBC bridge binary for ClickHouse, and a fat lib turns a database as an integration platform :)

> try to look to graalvm native binaries

Thanks @Slach. I'm not sure if it can address scenarios like adding/upgrading new driver without re-compilation.

**Comment by Slach** [comment]
> Thanks [@Slach](https://github.com/Slach). I'm not sure if it can address scenarios like adding/upgrading new driver without re-compilation.

I just think in containerization paradigm and distribution JDBC bridge together with clickhouse-server to avoid full JVM dependency.

**Comment by zhicwu** [comment]
> yep, you right, the same situation with ODBC

And we accept that? I respect the decision but to be honest it does not make sense to me. Also re-compiling won't resolve all problems. For example, the rare case we have is to setup two datasources pointing to same database server but using different version of driver for comparison and benchmark :)

**Comment by Slach** [comment]
> the rare case we have is to setup two datasources pointing to same database server but using different version of driver for comparison and benchmark :)

Rare cases require custom installation
you can create separate container images with separate JDBC drivers version and change `jdbc_bridge` config section to switch between versions
this is not clickhouse server and clickhouse-jdbc-bridge itself scope I think

**Comment by Slach** [comment]
> And we accept that? 
No me personally

But JDBC and ODBC drivers distributed separately and require custom installation steps and require bridge should know about new drivers is exists in system

What exactly you need? It looks like just external dependencies managing case. 
Which is your environment? Standalone server? VM? Containers?

**Comment by alexey-milovidov** [comment]
I don't see the work toward implementing an alternative - it means we cannot retire the JDBC bridge.
